From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrIvanPlays <pekov.ivan@abv.bg>
Date: Tue, 25 Jun 2019 20:20:49 +0300
Subject: [PATCH] Add option to choose between reconnect handlers

This patch adds json reconnect handler and
a option to choose between yaml and json
reconnect handlers.

diff --git a/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java b/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
index fc2257080e1efc3b5f8f80cfa795a2ec3356ca78..08c32f6ec3dfdc62102983067c47dd49466cf358 100644
--- a/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
+++ b/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
@@ -306,5 +306,12 @@ public interface ProxyConfig
      * @return should check for global op independently
      */
     boolean shouldCheckForGlobalOpIndependently();
+
+    /**
+     * Gets the reconnect handler type
+     *
+     * @return type of the reconnect handler
+     */
+    String getReconnectHandlerType();
     // IvanCord end
 }
diff --git a/proxy/src/main/java/com/mrivanplays/ivancord/JsonReconnectHandler.java b/proxy/src/main/java/com/mrivanplays/ivancord/JsonReconnectHandler.java
new file mode 100644
index 0000000000000000000000000000000000000000..2eb4531fe3f627b3dbf92cdc4ae02fa7a3aef921
--- /dev/null
+++ b/proxy/src/main/java/com/mrivanplays/ivancord/JsonReconnectHandler.java
@@ -0,0 +1,94 @@
+package com.mrivanplays.ivancord;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.logging.Level;
+
+import net.md_5.bungee.api.AbstractReconnectHandler;
+import net.md_5.bungee.api.ProxyServer;
+import net.md_5.bungee.api.config.ServerInfo;
+import net.md_5.bungee.api.connection.ProxiedPlayer;
+import net.md_5.bungee.config.Configuration;
+import net.md_5.bungee.config.ConfigurationProvider;
+import net.md_5.bungee.config.JsonConfiguration;
+
+public class JsonReconnectHandler extends AbstractReconnectHandler
+{
+
+    private Configuration configuration;
+    private File file;
+    private ConfigurationProvider provider;
+    private Map<UUID, String> data;
+
+    public JsonReconnectHandler()
+    {
+        this.file =  new File( "locations.json" );
+        this.provider = ConfigurationProvider.getProvider( JsonConfiguration.class );
+        if ( !file.exists() )
+        {
+            try
+            {
+                file.createNewFile();
+            } catch ( IOException e )
+            {
+            }
+        }
+        try
+        {
+            configuration = provider.load( file );
+        } catch ( IOException e )
+        {
+            file.renameTo( new File( "locations.json.old" ) );
+            ProxyServer.getInstance().getLogger().log( Level.WARNING, "Could not load reconnect locations, resetting them" );
+        }
+        data = new ConcurrentHashMap<>();
+        for ( String key : configuration.getKeys() )
+        {
+            String server = configuration.getString( key );
+            data.put( UUID.fromString( key ), server );
+        }
+    }
+
+    @Override
+    protected ServerInfo getStoredServer(ProxiedPlayer player)
+    {
+        return ProxyServer.getInstance().getServerInfo( data.get( player.getUniqueId() ) );
+    }
+
+    @Override
+    public void setServer(ProxiedPlayer player)
+    {
+        String server = player.getReconnectServer() != null ? player.getReconnectServer().getName() : player.getServer().getInfo().getName();
+        if ( !data.containsKey( player.getUniqueId() ) )
+        {
+            data.put( player.getUniqueId(), server );
+        } else
+        {
+            data.replace( player.getUniqueId(), server );
+        }
+    }
+
+    @Override
+    public void save()
+    {
+        for ( Map.Entry<UUID, String> entry : data.entrySet() )
+        {
+            configuration.set( entry.getKey().toString(), entry.getValue() );
+        }
+        try
+        {
+            provider.save( configuration, file );
+        } catch ( IOException e )
+        {
+            e.printStackTrace();
+        }
+    }
+
+    @Override
+    public void close()
+    {
+    }
+}
diff --git a/proxy/src/main/java/com/mrivanplays/ivancord/ReconnectHandlerType.java b/proxy/src/main/java/com/mrivanplays/ivancord/ReconnectHandlerType.java
new file mode 100644
index 0000000000000000000000000000000000000000..c9723674a638b75bba964ec0060848ce6fb8f200
--- /dev/null
+++ b/proxy/src/main/java/com/mrivanplays/ivancord/ReconnectHandlerType.java
@@ -0,0 +1,7 @@
+package com.mrivanplays.ivancord;
+
+public enum ReconnectHandlerType
+{
+    YAML,
+    JSON
+}
diff --git a/proxy/src/main/java/com/mrivanplays/ivancord/conf/IvanCordConfiguration.java b/proxy/src/main/java/com/mrivanplays/ivancord/conf/IvanCordConfiguration.java
index 6de414ca098ea7335e63b6c672f90e4e7240ca58..8da2ff6a2b590968422c23a456552223e2501ba7 100644
--- a/proxy/src/main/java/com/mrivanplays/ivancord/conf/IvanCordConfiguration.java
+++ b/proxy/src/main/java/com/mrivanplays/ivancord/conf/IvanCordConfiguration.java
@@ -2,6 +2,7 @@ package com.mrivanplays.ivancord.conf;
 
 import java.io.File;
 import java.util.logging.Level;
+import com.mrivanplays.ivancord.ReconnectHandlerType;
 import io.github.waterfallmc.waterfall.conf.WaterfallConfiguration;
 import lombok.Getter;
 import lombok.experimental.Accessors;
@@ -14,6 +15,7 @@ public class IvanCordConfiguration extends WaterfallConfiguration
 {
 
     private boolean kickToLobby = true;
+    private String reconnectHandlerType = ReconnectHandlerType.YAML.name();
 
     @Accessors(fluent = true)
     private boolean isTCPFastOpenEnabled = true;
@@ -32,6 +34,7 @@ public class IvanCordConfiguration extends WaterfallConfiguration
         isTCPFastOpenEnabled = config.getBoolean( "tcp_fast_open_enabled", isTCPFastOpenEnabled );
         getTCPFastOpenMode = setupTFO( config.getInt( "tcp_fast_open_mode", getTCPFastOpenMode ) );
         shouldCheckForGlobalOpIndependently = config.getBoolean( "should_check_for_global_op_independently", shouldCheckForGlobalOpIndependently );
+        reconnectHandlerType = setupReconnectHandlerType( config.getString( "reconnect_handler_type", reconnectHandlerType ) );
     }
 
     private int setupTFO(int incoming)
@@ -40,4 +43,16 @@ public class IvanCordConfiguration extends WaterfallConfiguration
         ProxyServer.getInstance().getLogger().log( Level.INFO, "TCP fast open mode is set to {0}", actual.name() );
         return actual.getValue();
     }
+
+    private String setupReconnectHandlerType(String incoming)
+    {
+        for ( ReconnectHandlerType type : ReconnectHandlerType.values() )
+        {
+            if ( type.name().toLowerCase().startsWith( incoming.toLowerCase() ) )
+            {
+                return type.name();
+            }
+        }
+        return ReconnectHandlerType.YAML.name();
+    }
 }
diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
index b7b06c7f0c44a48ebedcc5da155eb3c2f62b233b..39449556690a241c07f0961aac6f77cec6ae9005 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
@@ -96,6 +96,12 @@ import net.md_5.bungee.query.RemoteQuery;
 import net.md_5.bungee.scheduler.BungeeScheduler;
 import net.md_5.bungee.util.CaseInsensitiveMap;
 
+// IvanCord start
+import com.mrivanplays.ivancord.JsonReconnectHandler;
+import com.mrivanplays.ivancord.ReconnectHandlerType;
+import com.mrivanplays.ivancord.modules.YamlReconnectHandler;
+// IvanCord end
+
 /**
  * Main BungeeCord proxy class.
  */
@@ -317,6 +323,23 @@ public class BungeeCord extends ProxyServer
 
         isRunning = true;
 
+        // IvanCord start - setup reconnect handler
+        for ( ListenerInfo info : getConfig().getListeners() )
+        {
+            if ( !info.isForceDefault() && getReconnectHandler() == null )
+            {
+                if ( ReconnectHandlerType.valueOf( getConfig().getReconnectHandlerType() ) == ReconnectHandlerType.YAML )
+                {
+                    setReconnectHandler( new YamlReconnectHandler() );
+                } else
+                {
+                    setReconnectHandler( new JsonReconnectHandler() );
+                }
+                break;
+            }
+        }
+        // IvanCord end
+
         pluginManager.enablePlugins();
 
         if ( config.getThrottle() > 0 )
